parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -0,0 +1,169 @@
+              -- names were removed, we just need to delete from the index using a
+              -- prefix query, this would be faster than crawling to get all the deletes.
+              removedChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches removedChildBranch Branch.empty
+            That ca -> do
+              newChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches Branch.empty newChildBranch
+            These fromC toC
+              | Causal.valueHash fromC == Causal.valueHash toC -> do
+                  -- This child didn't change.
+                  pure Nothing
+              | otherwise -> do
+                  fromChildBranch <- Causal.value fromC
+                  toChildBranch <- Causal.value toC
+                  diffBranches fromChildBranch toChildBranch >>= \case
+                    Lens.Empty -> pure Nothing
+                    TreeDiff cfr -> pure . Just $ cfr
+        )
+  pure $ TreeDiff (defDiff :< childDiff)
+  where
+    diffMap :: forall ref. Ord ref => Map NameSegment (Map ref (m MdValues)) -> Map NameSegment (Map ref (m MdValues)) -> Map NameSegment (Diff ref)
+    diffMap l r =
+      Align.align l r
+        & fmap
+          ( \case
+              (This refs) -> (Diff {removals = Map.keysSet refs, adds = mempty})
+              (That refs) -> (Diff {removals = mempty, adds = Map.keysSet refs})
+              (These l' r') ->
+                let lRefs = Map.keysSet l'
+                    rRefs = Map.keysSet r'
+                 in (Diff {removals = lRefs `Set.difference` rRefs, adds = rRefs `Set.difference` lRefs})
+          )
+
+-- | Get a summary of all of the name adds and removals from a tree diff.
+--
+-- The provided name will be used as a prefix, and can be useful if diffing branches at a
+-- specific sub-tree, but you can pass 'Nothing' if you're diffing from the root.
+nameChanges ::
+  Maybe Name ->
+  TreeDiff ->
+  NameChanges
+nameChanges namePrefix (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< children)) =
+  let (termNameAdds, termNameRemovals) =
+        ( termDiffs
+            & ifoldMap \ns diff ->
+              let name = appendName ns
+               in (go name $ adds diff, go name $ removals diff)
+        )
+      (typeNameAdds, typeNameRemovals) =
+        ( typeDiffs
+            & ifoldMap \ns diff ->
+              let name = appendName ns
+               in (go name $ adds diff, go name $ removals diff)
+        )
+      childNameChanges =
+        ( children
+            & ifoldMap \ns childTree ->
+              nameChanges (Just $ appendName ns) (TreeDiff childTree)
+        )
+   in NameChanges {termNameAdds, termNameRemovals, typeNameAdds, typeNameRemovals} <> childNameChanges
+  where
+    appendName ns = maybe (Name.fromSegment . NameSegment.NameSegment . coerce $ ns) (`Lens.snoc` NameSegment.NameSegment (coerce ns)) namePrefix
```

Comment:
Maybe a type signature here? Why call `NameSegment` constructor but also call `coerce`? Maybe even use `case` instead of `maybe`.
```suggestion
    appendName :: NameSegment -> Name
    appendName ns = case namePrefix of
      Nothing -> Name.fromSegment ns
      Just prefix -> prefix `Lens.snoc` ns
```

Reply:
All of the 'coerce's are due to the v1 vs v2 NameSegment conversions 😢  I can probably just use `namesegment2to1` though.

Reply:
Ahh. Yeah `namesegment2to1` is probably been clearer, and we can also switch to v2 NameSegments across the board I guess if UCM doesn't mind using the same type as the database.  I guess as we get more concretely closer to the db representations in the CLI, this will be natural.

Reply:
I swapped to namesegment2to1 👍🏼